controller/postgres/deletion_protection.go

```diff
@@ -0,0 +1,128 @@
+package postgres
+
+import (
+	"context"
+	"encoding/json"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logging "sigs.k8s.io/controller-runtime/pkg/log"
+	"strconv"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	finalizerName  = "appcat.io/deletionProtection"
+	currentTimeKey = "now"
+)
+
+type jsonOp string
+
+const (
+	opRemove  jsonOp = "remove"
+	opAdd     jsonOp = "add"
+	opNone    jsonOp = "none"
+	opReplace jsonOp = "replace"
+)
+
+type jsonpatch struct {
+	Op    jsonOp `json:"op,omitempty"`
+	Path  string `json:"path,omitempty"`
+	Value string `json:"value,omitempty"`
+}
+
+func handle(ctx context.Context, inst client.Object, enabled bool, retention int) (client.Patch, error) {
+	log := logging.FromContext(ctx, "namespace", inst.GetNamespace(), "instance", inst.GetName())
+	op := opNone
+
+	if !enabled {
+		log.Info("DeletionProtection is not enabled, ensuring no finalizer set", "objectName", inst.GetName())
+		removed := controllerutil.RemoveFinalizer(inst, finalizerName)
+
+		if removed {
+			op = opRemove
+		}
+
+		return getPatchObjectFinalizer(log, inst, op)
+	}
+
+	if !controllerutil.ContainsFinalizer(inst, finalizerName) && inst.GetDeletionTimestamp() == nil {
+		added := controllerutil.AddFinalizer(inst, finalizerName)
+		if added {
+			log.Info("Added finalizer to the object", "objectName", inst.GetName())
+			op = opAdd
+			return getPatchObjectFinalizer(log, inst, op)
+		}
+	}
+
+	if inst.GetDeletionTimestamp() != nil {
+		op = checkRetention(ctx, inst, retention)
+	}
+
+	return getPatchObjectFinalizer(log, inst, op)
+}
```

Comment:
I had a hard time following this and I'm not completely sure this doesn't have any hidden bugs. I don't have a specific improvement suggestion so feel free to ignore this, but this just doesn't feel like the right abstraction

Reply:
I actually agree with this is not the best approach but I decided to keep what @Kidswiss started and just improve a bit the overall code. For instance the Patch operation was inside one of these functions and I did a refactoring to get it out and to it in the main reconciliation struct. This allowed me to skip passing client and some other objects around. If you have smth better in mind that would not take much time I'm open for change.

Reply:
I added multiple unit tests for these functions which should have caught almost all bugs.